### ISSUE TITLE
BroadcastException null socket

### ISSUE THREAD

**Issue by stoutZero** [beyondcode/laravel-websockets#261]
### TLDR
Enduser of my system will be making some sort of Order based on customer's request,  
that Order will generate notification to many users via broadcast.

Broadcast is queued, and handled by horizon.

But all broadcast related jobs failed, with this error in log file:  
`Illuminate\\Broadcasting\\BroadcastException(code: 0)`

Including `statistics-updated` event from websocket dashboard...

### Long and winding

- OS: `macOS 10.14.6`
- PHP Version: `PHP 7.2.17 (cli) (built: Apr 19 2019 00:49:48) ( NTS )`
- Package Versions:
  - `laravel/framework`: `v5.8.35`
  - `beyondcode/laravel-websockets`: `1.3.0`
  - `pusher/pusher-php-server`: `v4.0.0`

- My config:
  <details>
    <summary>.env</summary>
  
  ```ini
  APP_NAME="<SOME_VALUE>"
  APP_ENV=local
  APP_KEY=base64:<SOME_VALUE>
  APP_DEBUG=true
  APP_EDITOR=phpstorm
  APP_HTTPS=false
  APP_URL=http://localhost
  APP_TIMEZONE="Asia/Jakarta"

  AWS_ACCESS_KEY_ID=
  AWS_SECRET_ACCESS_KEY=
  AWS_DEFAULT_REGION=us-east-1
  AWS_BUCKET=

  BROADCAST_DRIVER=pusher

  CACHE_DRIVER=file

  # omitted...

  LARAVEL_WEBSOCKETS_SSL_LOCAL_CERT=
  LARAVEL_WEBSOCKETS_SSL_LOCAL_PK=
  LARAVEL_WEBSOCKETS_SSL_PASSPHRASE=
  LARAVEL_WEBSOCKETS_PORT=6123

  LOG_CHANNEL=stack

  MAIL_DRIVER=smtp
  MAIL_HOST=smtp.mailtrap.io
  MAIL_PORT=2525
  MAIL_USERNAME=null
  MAIL_PASSWORD=null
  MAIL_ENCRYPTION=null

  PUSHER_APP_ID=<SOME_VALUE>
  PUSHER_APP_KEY=local-dev
  PUSHER_APP_SECRET=super-secret-key
  PUSHER_APP_CLUSTER=mt1
  MIX_PUSHER_APP_KEY="${PUSHER_APP_KEY}"
  MIX_PUSHER_APP_CLUSTER="${PUSHER_APP_CLUSTER}"

  QUEUE_CONNECTION=redis

  REDIS_HOST=127.0.0.1
  REDIS_PASSWORD=null
  REDIS_PORT=6379

  SESSION_DRIVER=file
  SESSION_LIFETIME=120

  # this should only have values on prod servers, or when testing temporarily
  # sentry notification email can be quite noisy when debugging
  SENTRY_LARAVEL_DSN=

  SENTRY_BROWSER_SRC=
  SENTRY_BROWSER_INTEGRITY=
  SENTRY_BROWSER_CROSSORIGIN=
  MIX_SENTRY_AUTH_TOKEN="<SOME_VALUE>"

  TELESCOPE_CONNECTION=telescope
  TELESCOPE_DRIVER=database
  ```
  </details>

  <details>
    <summary>websockets.php</summary>
    
  ```php
  return [
    // unchanged...
    
    'dashboard' => [
        'port' => env('LARAVEL_WEBSOCKETS_PORT', 6123),
    ],

    // unchanged...
    
    'apps' => [
        [
            'id' => env('PUSHER_APP_ID'),
            'name' => env('APP_NAME'),
            'key' => env('PUSHER_APP_KEY'),
            'secret' => env('PUSHER_APP_SECRET'),
            'path' => env('PUSHER_APP_PATH'),
            'capacity' => null,
            'enable_client_messages' => false,
            'enable_statistics' => true,
        ],
    ],

    // unchanged...
    
    'path' => 'ws-dash',
  ];
  ```
  </details>
  
  <details>
    <summary>broadcasting.php</summary>
    
  ```php
  return [
    // unchanged...
    
    'pusher' => [
        'driver' => 'pusher',
        'app_id' => env('PUSHER_APP_ID'),
        'key' => env('PUSHER_APP_KEY'),
        'secret' => env('PUSHER_APP_SECRET'),
        'options' => [
            'cluster' => env('PUSHER_APP_CLUSTER'),
            'encrypted' => env('APP_ENV') === 'production' ? true : false,
            'host' => '127.0.0.1',
            'port' => env('LARAVEL_WEBSOCKETS_PORT', 6123),
            'socket' => env('LARAVEL_WEBSOCKETS_PORT', 6123),
            'scheme' => !env('APP_HTTPS') ? 'http' : 'https',
        ],
    ],

    // unchanged...
  ];
  ```
  </details>

- Error message:
  <details>
    <summary>Click to view Error message</summary>

  ```
  [Y-m-d H:i:s] local.ERROR:  {"exception":"[object] (Illuminate\\Broadcasting\\BroadcastException(code: 0):  at <PROJECT>/vendor/laravel/framework/src/Illuminate/Broadcasting/Broadcasters/PusherBroadcaster.php:123)
  [stacktrace]
  #0 <PROJECT>/vendor/laravel/framework/src/Illuminate/Broadcasting/BroadcastEvent.php(48): Illuminate\\Broadcasting\\Broadcasters\\PusherBroadcaster->broadcast(Array, 'statistics-upda...', Array)
  #1 [internal function]: Illuminate\\Broadcasting\\BroadcastEvent->handle(Object(Illuminate\\Broadcasting\\Broadcasters\\PusherBroadcaster))
  #2 <PROJECT>/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(32): call_user_func_array(Array, Array)
  #3 <PROJECT>/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(90): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
  #4 <PROJECT>/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(34): Illuminate\\Container\\BoundMethod::callBoundMethod(Object(Illuminate\\Foundation\\Application), Array, Object(Closure))
  #5 <PROJECT>/vendor/laravel/framework/src/Illuminate/Container/Container.php(576): Illuminate\\Container\\BoundMethod::call(Object(Illuminate\\Foundation\\Application), Array, Array, NULL)
  #6 <PROJECT>/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(94): Illuminate\\Container\\Container->call(Array)
  #7 <PROJECT>/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Bus\\Dispatcher->Illuminate\\Bus\\{closure}(Object(Illuminate\\Broadcasting\\BroadcastEvent))
  #8 <PROJECT>/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(104): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Broadcasting\\BroadcastEvent))
  #9 <PROJECT>/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(98): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
  #10 <PROJECT>/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(49): Illuminate\\Bus\\Dispatcher->dispatchNow(Object(Illuminate\\Broadcasting\\BroadcastEvent), false)
  #11 <PROJECT>/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(88): Illuminate\\Queue\\CallQueuedHandler->call(Object(Illuminate\\Queue\\Jobs\\RedisJob), Array)
  #12 <PROJECT>/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(337): Illuminate\\Queue\\Jobs\\Job->fire()
  #13 <PROJECT>/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(283): Illuminate\\Queue\\Worker->process('redis', Object(Illuminate\\Queue\\Jobs\\RedisJob), Object(Illuminate\\Queue\\WorkerOptions))
  #14 <PROJECT>/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(118): Illuminate\\Queue\\Worker->runJob(Object(Illuminate\\Queue\\Jobs\\RedisJob), 'redis', Object(Illuminate\\Queue\\WorkerOptions))
  #15 <PROJECT>/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(102): Illuminate\\Queue\\Worker->daemon('redis', 'default', Object(Illuminate\\Queue\\WorkerOptions))
  #16 <PROJECT>/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(86): Illuminate\\Queue\\Console\\WorkCommand->runWorker('redis', 'default')
  #17 <PROJECT>/vendor/laravel/horizon/src/Console/WorkCommand.php(46): Illuminate\\Queue\\Console\\WorkCommand->handle()
  #18 [internal function]: Laravel\\Horizon\\Console\\WorkCommand->handle()
  #19 <PROJECT>/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(32): call_user_func_array(Array, Array)
  #20 <PROJECT>/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(90): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
  #21 <PROJECT>/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(34): Illuminate\\Container\\BoundMethod::callBoundMethod(Object(Illuminate\\Foundation\\Application), Array, Object(Closure))
  #22 <PROJECT>/vendor/laravel/framework/src/Illuminate/Container/Container.php(576): Illuminate\\Container\\BoundMethod::call(Object(Illuminate\\Foundation\\Application), Array, Array, NULL)
  #23 <PROJECT>/vendor/laravel/framework/src/Illuminate/Console/Command.php(183): Illuminate\\Container\\Container->call(Array)
  #24 <PROJECT>/vendor/symfony/console/Command/Command.php(255): Illuminate\\Console\\Command->execute(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Illuminate\\Console\\OutputStyle))
  #25 <PROJECT>/vendor/laravel/framework/src/Illuminate/Console/Command.php(170): Symfony\\Component\\Console\\Command\\Command->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Illuminate\\Console\\OutputStyle))
  #26 <PROJECT>/vendor/symfony/console/Application.php(915): Illuminate\\Console\\Command->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
  #27 <PROJECT>/vendor/symfony/console/Application.php(272): Symfony\\Component\\Console\\Application->doRunCommand(Object(Laravel\\Horizon\\Console\\WorkCommand), Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
  #28 <PROJECT>/vendor/symfony/console/Application.php(148): Symfony\\Component\\Console\\Application->doRun(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
  #29 <PROJECT>/vendor/laravel/framework/src/Illuminate/Console/Application.php(90): Symfony\\Component\\Console\\Application->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
  #30 <PROJECT>/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(133): Illuminate\\Console\\Application->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
  #31 <PROJECT>/artisan(37): Illuminate\\Foundation\\Console\\Kernel->handle(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
  #32 {main}
  "} 
  ```
  </details>
  
Websocket is being run using this command:  
`php artisan -vvv websockets:serve --debug --host=127.0.0.1 --port=6123`

or

`php artisan -vvv websockets:serve --debug --port=6123`

Broadcast is queued and handled by **horizon**.

Upon changing `@broadcast` method on `<PROJECT>/vendor/laravel/framework/src/Illuminate/Broadcasting/Broadcasters/PusherBroadcaster.php:123` into:

  ```php
  public function broadcast(array $channels, $event, array $payload = [])
  {
      // here
      \Log::info(compact('payload'));

      $socket = Arr::pull($payload, 'socket');

      $response = $this->pusher->trigger(
          $this->formatChannels($channels), $event, $payload, $socket, true
      );

      // and there
      \Log::info(compact('response'));

      if ((is_array($response) && $response['status'] >= 200 && $response['status'] <= 299)
          || $response === true) {
          return;
      }

      throw new BroadcastException(
          is_bool($response) ? 'Failed to connect to Pusher.' : $response['body']
      );
  }
  ```

I got these on `laravel.log`
  ```log
  [Y-m-d H:i:s] local.INFO: array (
    'payload' => 
    array (
      'time' => 'Y-m-d H:i:s',
      'app_id' => '<APP_ID>',
      'peak_connection_count' => 1,
      'websocket_message_count' => 1,
      'api_message_count' => 0,
      'socket' => NULL, // this is always null
    ),
  )  
  [Y-m-d H:i:s] local.INFO: array (
    'response' => 
    array (
      'body' => false,
      'status' => 0,
    ),
  )  
  ```

Nothing is running on port `6123`, and this is `statistics-updated` event that  
`beyondcode/laravel-websockets` itself is generating...

Looking at codes in `Pusher\Pusher` class,  
`@validate_socket_id` method seems to allow null socket, which mean it won't throw  
any exception...
`@validate_socket_id` was called inside `@trigger` method  
which should return either `boolean` or an array in this structure:

```php
$response = [
  'status' => 200 // valid value: 200 <= n => 299
];
```

but looking at what's in the log file, `status` is `0` (zero)...

Right now I can't debug any deeper, sometime tomorrow maybe...

Any help will be greatly appreciated...

_EDIT (2019-10-02 00:05 GMT+7): Added TLDR explanation_  
_EDIT (2019-10-05 08:30 GMT+7): Added relevant `.env` content_

**Comment by stri8ed** [comment]
What is the problem you are trying to solve?

**Comment by stoutZero** [comment]
> What is the problem you are trying to solve?

Uhhh... this: `local.ERROR:  {"exception":"[object] (Illuminate\\Broadcasting\\BroadcastException(code: 0)`

**Comment by indigoram89** [comment]
I have the same error :(

**Comment by canhkieu** [comment]
Could you send .env content. Maybe something configs not valid.

**Comment by stoutZero** [comment]
> Could you send .env content. Maybe something configs not valid. Or I recommend use default port 6001 for socket connect.

Updated, take a look

**Comment by swash13** [comment]
I have the same error, did you found how to solve this? I debugged it and it fails on curl execution.
Here is my code:
Javascript:
`window.Echo = new Echo({
    broadcaster: 'pusher',
    key: process.env.MIX_PUSHER_APP_KEY,
    wsHost: window.location.hostname,
    wsPort: 6001,
    disableStats: true,
});`

broadcasting.php
`'pusher' => [
            'driver' => 'pusher',
            'key' => env('PUSHER_APP_KEY'),
            'secret' => env('PUSHER_APP_SECRET'),
            'app_id' => env('PUSHER_APP_ID'),
            'options' => [
                'cluster' => env('PUSHER_APP_CLUSTER'),
                'encrypted' => env('PUSHER_APP_ENCRYPTED'),
                'host' => env( 'PUSHER_APP_HOST', '127.0.0.1'),
                'port' => env( 'PUSHER_APP_PORT', 6001),
                'scheme' => env( 'PUSHER_APP_SCHEME', 'http'),
            ],
        ],`

.env file:
`BROADCAST_DRIVER=pusher

PUSHER_APP_ID=some-app-id
PUSHER_APP_KEY=some-app-key
PUSHER_APP_SECRET=some-secret
PUSHER_APP_CLUSTER=mt1
PUSHER_APP_PORT=6001
PUSHER_APP_HOST=product-management-app.test
PUSHER_APP_ENCRYPTED=false
PUSHER_APP_SCHEME=http`

**Comment by swash13** [comment]
Guys, i found the issue. If you're on localhost, try set in `broadcasting.php` file PUSHER_APP_HOST to 127.0.0.1 instead of your domain that it points to and it will work.
If you're using forge - practice is the same. For javascript it stays the same `window.location.hostname`

**Comment by stoutZero** [comment]
> Guys, i found the issue. If you're on localhost, try set in `broadcasting.php` file PUSHER_APP_HOST to 127.0.0.1 instead of your domain that it points to and it will work.
> If you're using forge - practice is the same, set `127.0.0.1`. For javascript it stays the same `window.location.hostname`

This is the first time i'm seeing `PUSHER_APP_HOST` env... I almost always see `127.0.0.1` anywhere...

**Comment by stoutZero** [comment]
Haaa... forget it...

**Comment by indigoram89** [comment]
In my case, it will be too many connections. Some tuning of the server solves the problem.

**Comment by Cello93** [comment]
> En mi caso, habrá demasiadas conexiones. Algunos ajustes del servidor resuelven el problema.

what adjustment did you make?

**Comment by tonz341** [comment]
hi we have hosted our socket server on digital ocean.
currently it is using domain url as PUSHER_APP_HOST  like socket.example.com

and theres error popping out on our laravel log
 {"exception":"[object] (Illuminate\\Broadcasting\\BroadcastException(code: 0)

can i change to 127.0.0.1 instead of domain name to remove the error?

**Comment by gbelot2003** [comment]
Incredible, at the end of the day, the problem is solved by changing the host, since my host starts in workspace, the host option in broadcasting should point to it.
` 'options' => [
                'cluster' => 'eu',
                'useTLS' => false,
                'encrypted' => false,
                'host' => 'workspace',
                'port' => 6001,
                'scheme' => 'http',
                'curl_options' => [
                    CURLOPT_SSL_VERIFYHOST => 0,
                    CURLOPT_SSL_VERIFYPEER => 0,
                ],
            ],`
